Implement ClusterRole, RoleBinding, and ClusterRoleBinding syncing

mattkelly opened the pull request:
### Description

 #### What does this pull request accomplish?

Implement ClusterRole, RoleBinding, and ClusterRoleBinding syncing based on the following logic previously discussed:

1. Sync all roles (which include rules) from sideload route to Kubernetes ClusterRoles (`name` == cloud role ID)
2. Sync all role bindings for each role to ClusterRoleBindings or RoleBindings  depending on if `namespace` is set

- TeamBinding corresponds to Kubernetes `kind = Group`, `subject` is `team_id`
- UserBinding corresponds to Kubernetes `kind = User`, `subject` is `user_id`

 #### What issue(s) does this fix?
* Closes #63

 #### Additional Considerations

 ### Testing

In progress - currently noticing a few weird things which I've inquired about in Slack

 ### Dependencies
 None

mattkelly commented on pkg/coordinator/authorization_role_binding_controller.go:
```diff
@@ -0,0 +1,334 @@
+
+	roleNamespace := authorizationRoleBinding.Namespace
+	if roleNamespace == "" {
+		// No namespace, so this should be synced to a ClusterRoleBinding
```
These blocks could probably be pulled out into helpers

mattkelly replied:
Fiddled with this - it's not really much cleaner to pull these out, so I'm leaving as-is for now.

mattkelly commented on pkg/coordinator/authorization_role_binding_controller.go:
```diff
@@ -0,0 +1,334 @@
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   authRoleBinding.Spec.ID,
+			Labels: constants.BuildContainershipLabelMap(nil),
+			OwnerReferences: []metav1.OwnerReference{
```
Consider adding a helper to create this

nicktate commented on pkg/coordinator/authorization_role_binding_controller.go:
```diff
@@ -0,0 +1,334 @@
+)
+
+// AuthorizationRoleBindingController is the controller implementation for the containership
+// upgrading clusters to a users desired kubernetes version
```
should this be more about rbac than cluster upgrades?

mattkelly replied:
Probably

nicktate commented on pkg/coordinator/authorization_role_controller.go:
```diff
@@ -0,0 +1,261 @@
+)
+
+// AuthorizationRoleController is the controller implementation for the containership
+// upgrading clusters to a users desired kubernetes version
```
lets update this comment to refer to syncing roles / rules

nicktate commented on pkg/coordinator/authorization_role_controller.go:
```diff
@@ -0,0 +1,261 @@
+			Verbs:         authRule.Verbs,
+			APIGroups:     authRule.APIGroups,
+			Resources:     authRule.Resources,
+			ResourceNames: authRule.ResourceNames,
```
It's not yet part of the actual cloud auth rule, but we are going to add the following key to the `authorizationRule` 

`nonResourceURLs` : https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.13/#policyrule-v1beta1-rbac-authorization-k8s-io

We should add it in now so when it is available, it is already good to go

nicktate replied:
It's going to be set as `non_resource_urls` on the cloud api object

mattkelly commented:
Added non-resource URL support, fixed a bug where I was referencing the wrong namespace on binding creation, and addressed comments. Should be good for another look.

mattkelly commented:
Re-tested again this morning with what looks to be a cleaned up stage environment. Looks good to me.
